### DISCUSSION TITLE
Are Eclipse variables supported in target.platform

### DISCUSSION THREAD

**Discussion by SimonJohnCastle** [eclipse-tycho/tycho#550] [Q&A] [answered]
From the Eclipse target platform editor I can set up relative paths to plugin directories, this also appears to work when building with tycho.
When I build the target platform pom directly this work, but when I build from a parent it looks like it uses the ${project_loc} from the parent, so cannot find the plugin directory.
For the time being I can work round this by creating symbolic links in the filesystem (at least on linux).
Should this work or is it a bug ?
This becomes an issue when you checkout into a different location or user.

![image](https://user-images.githubusercontent.com/51445041/150532804-538736e7-bc4a-464b-8398-c975b4433baa.png)

**Comment by laeubi** [comment, marked as answer]
The `${project_loc}` requires the name of a project as a "basedirectory" as otherwise the current selection is used, see [this comment](https://bugs.eclipse.org/bugs/show_bug.cgi?id=578119#c5) for example.

Beside from that, Tycho supports the following eclipse variables:

- `system_property`
- `project_loc`
- `env_var`

**Reply by SimonJohnCastle** [reply]
Thanks for the reply, this looks to work but I dont fully understand whats going on and how it works.
For the first project, where the sub projects are all side by side including the parent and the external pluginFolder

`${project_loc:/targetProjectName}/../pluginsFolder`

This works in Eclipse and Maven and fits with the [description](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.user%2Fconcepts%2Fconcepts-exttools.htm) as I understand it.

However I have another project where the slash after the colon has to be removed

`${project_loc:targetProjectName}/../pluginsFolder`

In this case the parent is directly above the sub projects.

Its not clear to me when and why to use the leading slash, but I do have a working build.
Thank you

**Reply by laeubi** [reply]
Its all a bit a matter of how you name projects and alike, e.g. tycho tries to guess what reactor project matching a given name. In general it shouldn't matter if there is a slash or not (at least for tycho), you just need to make sure that there are no additional segments inside the brackets as this is not supported by tycho itself.

If you could provide a sample projects where it fails, it might be easier to guess why.

**Comment by ffays** [comment]
Neither system_property nor env_var variables are currently working in target platform definition on Mac OS:

```
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Invalid syntax in target definition ...
...
Caused by: org.eclipse.tycho.p2.target.facade.TargetDefinitionSyntaxException: invalid URI
...
Caused by: java.net.URISyntaxException: Illegal character in authority at index 7: file://${env_var:HOME}/git/myproject
...
```
```
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Invalid syntax in target definition ...
...
Caused by: org.eclipse.tycho.p2.target.facade.TargetDefinitionSyntaxException: invalid URI
...
Caused by: java.net.URISyntaxException: Illegal character in authority at index 7: file://${system_property:user.home}/git/myproject
...
```

**Reply by laeubi** [reply]
Please share a full project / target file or provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue.

**Comment by ffays** [comment]
Hereby a full project that demonstrate the issue: [tycho.issue550.zip](https://github.com/eclipse-tycho/tycho/files/10488812/tycho.issue550.zip)

Deploy the project in your `${HOME}/git` (unix) or `%USERPROFILE%\git` (windows)

Build: `cd ~/git/tycho.issue550 && mvn clean && mvn -f tycho.issue550.update/pom.xml package && mvn package` (unix)

Note the update site has to be build prior to build whole project as there is a dependency in target platform referring to a local plugin from the update site.

Could you please have a look?

**Reply by laeubi** [reply]
You are using variable in the `location` attribute of an updatesite what is not supported by PDE (and therefore not by Tycho) see:
- https://github.com/eclipse-pde/eclipse.pde/issues/204

If you don't care about IDE support, you can either define your repo in the `pom.xml` repository section or using mirrors in `settings.xml`, in both cases you can use whatever maven support at these places.

**Reply by cpfeiffer** [reply]
PDE actually does support variables in the location attribute, maybe not as sophisticated as described in https://github.com/eclipse-pde/eclipse.pde/issues/204

This simple form works in both PDE and in Tycho, however I'm not sure where to set `my.dir` for Tycho. I managed to set it in `.mvn/maven.config` via `-Dmy.dir=...`, but that's is not really helpful. A property from pom.xml would be more useful.
```
<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
		<repository location="file:${system_property:my.dir}/extra-deps"/>
</location>
```

**Reply by laeubi** [reply]
> This simple form works in both PDE and in Tycho, however I'm not sure where to set `my.dir` for Tycho. I managed to set it in `.mvn/maven.config` via `-Dmy.dir=...`, but that's is not really helpful.

I'm not sure what else is expected, e.g. in PDE you would need to set it in `eclipse.ini` so `maven.config` is the somehow equivalent...

> A property from pom.xml would be more useful.

Even though people often want to do that (can't we simply use X from project/pom/parent/...) this is not easily possible because a target is not tied to a project at all. It can be a file or even a remote resource and usually all projects share the same target file.

**Reply by cpfeiffer** [reply]
I see your point, and agree that one should be conscious where to use such variables in targets. But eclipse is pretty flexible concerning variable values, it can even pop up a dialog and ask the user for the desired value. So taking a value from a pom property doesn't seem far-fetched.

**Reply by merks** [reply]
FYI, there are places that I use the string substitutions for a similar purpose:

<img width="587" height="281" alt="image" src="https://github.com/user-attachments/assets/09001f04-f0ad-46f3-bac6-f36b7da43f6a" />

E.g., here I can use either a system property value or use a string substitution, if available.

https://github.com/eclipse-cbi/p2repo-aggregator/blob/62ea19d709a22164b0d4af9d35bcca502f555e19/plugins/org.eclipse.cbi.p2repo.util/src/org/eclipse/cbi/p2repo/util/StringUtils.java#L28-L70

In my client project, I wished that relative paths would work instead of the current approach where the project needs to be in the reactor...

**Reply by laeubi** [reply]
> I see your point, and agree that one should be conscious where to use such variables in targets. But eclipse is pretty flexible concerning variable values, it can even pop up a dialog and ask the user for the desired value. So taking a value from a pom property doesn't seem far-fetched.

As said you have 50 project using a target fetched from a http server, what "pom properties" should it use? So Tycho already supports a lot of things (including system properties) so I'm a bit confused now one expect "more flexible" support than that? A system property can neither change during a maven build (or more formally should not) nor in the IDE.

If you need maximum maven property expansion support you can define a target platform inline in the `pom.xml` see https://tycho.eclipseprojects.io/doc/latest/target-platform-configuration/target-platform-configuration-mojo.html#target .. of course than again it comes to that this is NOT supported in the IDE.

So the general contract here is, we support what can be supported in **both** worlds consistently in _target files_ (!) ... beside that it supports some extensions for defining _target contents_ (!) that is not supported in PDE at all.

> In my client project, I wished that relative paths would work instead of the current approach where the project needs to be in the reactor...

See 
- https://github.com/eclipse-pde/eclipse.pde/issues/169
- https://github.com/eclipse-pde/eclipse.pde/issues/204

As Tycho is already quite flexible and already *can* do a lot more it is more that firs PDE need to support it and then we can add support to Tycho.

**Reply by cpfeiffer** [reply]
If I build 50 projects in eclipse, I can use a single target platform that may contain a `system_property` variable and set the actual system property value in `eclipse.ini`, e.g. via oomph setup.

In Maven, one _could_ set the property in `pom.xml`. Of course, one could break things and set the property differently for every of the 50 plugins, but why? The intention would be to set it just once in the parent pom to be the same value for all 50 plugins.

That would be pretty consistent to the IDE build, IMHO.

**Reply by phermsdorf** [reply]
Can you elaborate a bit on your use case for this? Which problem would solve that for you?

**Reply by cpfeiffer** [reply]
Imagine you have two different builds, the first produces some artifacts locally, the second wants to consume them. You would create a target platform for the second build and reference the location of the locally produced artifacts. You can't hardcode the location, so a system property would be handy.

**Reply by laeubi** [reply]
> In Maven, one _could_ set the property in `pom.xml`.

No you can't. in `pom.xml `you can set _user-properties_ (!) that ar scoped to the current project. Actually you can only set *real* system properties in `jvm.config` but `-D` in `maven.config` works equally well because for the session maven merges system, settings and user properties.

So I really don't get why `maven.config` is not enough (and I actually using something similar in my setups) having them defined in parent pom would just add another level of complexity here because now even project inheritance comes into play (and not all projects must use the same parent (version).

So from my point of view you are trying to solve that goes far beyond a placeholder in a target file... if you need to share the same target you better deploy the target file itself as an artifact and then reference that (maybe with the same version as your parent pom).

**Reply by cpfeiffer** [reply]
>> In Maven, one _could_ set the property in pom.xml.
> No you can't.

Well, that's exactly what we're talking about. It doesn't work with properties defined in `pom.xml` (yet).

> So I really don't get why maven.config is not enough

Because it is usually in version control, so you can't have user-specific paths in there.

**Reply by laeubi** [reply]
> Well, that's exactly what we're talking about. It doesn't work with properties defined in pom.xml (yet).

We are spinning a bit in circles, again `pom.xml` has _NOTHING_ to do with target file... target must even not be a file or located in any folder where a pom is located.

Then variable you use here is `system_property` --> this is the plain java concept you get when you start the JVM and it works like this in PDE + Maven... so if you start your maven call with `mvn clean verify` (or whatever) do it with `mvn clean verify -DmySystemproperty=X`  and it works.

If you need **maven property (!)** this is only supported in `pom.xml` context not in `target-file` and you can use any `<configuration>` option that can be a `<url>` or `<location>` to or you can use `<repository>` and there you can use whatever maven properties you want.

**Reply by cpfeiffer** [reply]
You're absolutely right. I was just mentioning that this is suboptimal from the user's point of view. If I configure the target platform in pom.xml, I would also like to configure the properties so that it can be resolved. System properties were the best I could find, but hard to provide; pom properties were the best non-exact match.
